src/compiler/checker.ts

```diff
@@ -3605,10 +3604,18 @@ module ts {
             if (!type) {
                 type = unionTypes[id] = <UnionType>createObjectType(TypeFlags.Union | getWideningFlagsOfTypes(sortedTypes));
                 type.types = sortedTypes;
+                type.reducedType = noSubtypeReduction ? undefined : type;

```

Comment:
Could we accomplish this in a simpler manner by just having getUnionType always do subtype reduction?  Would that then ensure that all later parts of the checker see the union as consistently reduced?